src/vs/workbench/contrib/notebook/browser/view/renderers/webviewPreloads.ts

```diff
@@ -789,51 +844,88 @@ async function webviewPreloads(style: PreloadStyles, rendererData: readonly Webv
+			private readonly loadDependency: (id: string) => Promise<void>,
+		) { }
+
+		private _loadPromise: Promise<RendererApi> | undefined;
+		private _api: RendererApi | undefined;
+
+		public get api() { return this._api; }
+
+		public load(): Promise<RendererApi | undefined> {
+			if (!this._loadPromise) {
+				this._loadPromise = Promise.all(this.data.dependencies.map(dependencyId => this.loadDependency(dependencyId)))
+					.then(() => runRenderScript(this.data.entrypoint, this.data.id))
+					.then(module => {
+						if (module) {
+							const api = module.activate(createRendererContext(this.data.id));
+							this._api = api;
+							return api;
+						}
+						return undefined;
+					});
+			}
+			return this._loadPromise;
+		}
 	}
 
-	const markupRenderers = new class {
+	const renderers = new class {
 
-		private readonly mimeTypesToRenderers = new Map<string, {
-			load: () => Promise<MarkupRenderer>;
-		}>();
+		private readonly _renderers = new Map</* id */ string, Renderer>();
 
 		constructor() {
 			for (const renderer of rendererData) {
-				let loadPromise: Promise<MarkupRenderer> | undefined;
+				this._renderers.set(renderer.id, new Renderer(renderer, async (dependencyId) => {
+					const parent = this._renderers.get(dependencyId);
+					if (!parent) {
+						throw new Error(`Could not find renderer dependency: ${dependencyId}`);
+					}
+					await parent.load();
+				}));
+			}
+		}
 
-				const entry = {
-					load: () => {
-						if (!loadPromise) {
-							loadPromise = __import(renderer.entrypoint).then(module => {
-								return module.activate({ dependencies: renderer.dependencies });
-							});
-						}
-						return loadPromise;
-					},
-					renderer: undefined,
-				};
+		public getRenderer(id: string): RendererApi | undefined {
+			return this._renderers.get(id)?.api;
+		}
 
-				for (const mime of renderer.mimeTypes || []) {
-					if (!this.mimeTypesToRenderers.has(mime)) {
-						this.mimeTypesToRenderers.set(mime, entry);
-					}
-				}
+		public load(id: string) {
+			const renderer = this._renderers.get(id);
+			if (!renderer) {
+				throw new Error('Could not find renderer');
 			}
+
+			return renderer.load();
 		}
 
-		async renderMarkdown(element: HTMLElement, content: string): Promise<void> {
-			const entry = this.mimeTypesToRenderers.get('text/markdown');
-			if (!entry) {
+		public async renderMarkdown(id: string, element: HTMLElement, content: string): Promise<void> {
+			const markdownRenderers = Array.from(this._renderers.values())
+				.filter(renderer => renderer.data.mimeTypes.includes('text/markdown'));
+
+			if (!markdownRenderers.length) {
 				throw new Error('Could not find renderer');
 			}
-			const renderer = await entry.load();
-			renderer.renderMarkup({ element, content });
+
+			await Promise.all(markdownRenderers.map(x => x.load()));
+
+			const renderer = Array.from(this._renderers.values()).find(x => x.data.mimeTypes.includes('text/markdown'));
```

Comment:
This seems like it's a duplicated lookup where we could use `markdownRenderers[0]`